blas/impl/KokkosBlas1_axpby_unification_attempt_traits.hpp

```diff
@@ -0,0 +1,955 @@
+          ": if Y is not complex, then A, X and B cannot be complex");
+    }
+
+    // ******************************************************************
+    // Check 2/6: YMV is valid
+    // ******************************************************************
+    static_assert(
+        Kokkos::is_view<YMV>::value,
+        "KokkosBlas::Impl::AxpbyUnificationAttemptTraits::performChecks()"
+        ": Y is not a Kokkos::View.");
+    static_assert(
+        std::is_same<typename YMV::value_type,
+                     typename YMV::non_const_value_type>::value,
+        "KokkosBlas::Impl::AxpbyUnificationAttemptTraits::performChecks()"
+        ": Y is const.  It must be nonconst, "
+        "because it is an output argument "
+        "(we must be able to write to its entries).");
+    static_assert(
+        Kokkos::SpaceAccessibility<tExecSpace,
+                                   typename YMV::memory_space>::accessible,
+        "KokkosBlas::Impl::AxpbyUnificationAttemptTraits::performChecks()"
+        ": XMV must be accessible from tExecSpace");
+
+    // ******************************************************************
+    // Check 3/6: XMV is valid
+    // ******************************************************************
+    static_assert(
+        Kokkos::is_view<XMV>::value,
+        "KokkosBlas::Impl::AxpbyUnificationAttemptTraits::performChecks()"
+        ": X is not a Kokkos::View.");
+    static_assert(
+        Kokkos::SpaceAccessibility<tExecSpace,
+                                   typename XMV::memory_space>::accessible,
+        "KokkosBlas::Impl::AxpbyUnificationAttemptTraits::performChecks()"
+        ": XMV must be accessible from tExecSpace");
+
+    if constexpr (xyRank1Case) {
+      if (X.extent(0) == Y.extent(0)) {
+        // Ok
+      } else {
+        std::ostringstream msg;
+        msg << "KokkosBlas::Impl::AxpbyUnificationAttemptTraits::performChecks("
+               ")"
+            << ", invalid rank-1 X extent"
+            << ": X.extent(0) = " << X.extent(0)
+            << ", Y.extent(0) = " << Y.extent(0);
+        KokkosKernels::Impl::throw_runtime_exception(msg.str());
+      }
+    } else {
+      if ((X.extent(0) == Y.extent(0)) && (X.extent(1) == Y.extent(1))) {
+        // Ok
+      } else {
+        std::ostringstream msg;
+        msg << "KokkosBlas::Impl::AxpbyUnificationAttemptTraits::performChecks("
+               ")"
+            << ", invalid rank-2 X extents"
+            << ": X.extent(0) = " << X.extent(0)
+            << ", X.extent(1) = " << X.extent(1)
+            << ", Y.extent(0) = " << Y.extent(0)
+            << ", Y.extent(1) = " << Y.extent(1);
+        KokkosKernels::Impl::throw_runtime_exception(msg.str());
+      }
+    }
+
+    // ******************************************************************
+    // Check 4/6: AV is valid
+    // ******************************************************************
+    static_assert(
```

Comment:
This statement can be reduced to: `(((a_is_scalar != a_is_r0) != a_is_r1s) != a_is_r1d)`, please double check the truth table.

Reply:
Thanks, Luc, but here I prefer to keep the way it is. Yes, it is much "larger" code, but I think the understanding is pretty straightforward the way it is now.